eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/EventHubSender.java

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.eventhubs;
+
+import com.azure.core.implementation.logging.ServiceLogger;
+import com.azure.eventhubs.implementation.ClientConstants;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+/**
+ * This sender class is a logical representation of sending events to Event Hubs.
+ *
+ * @see EventHubClient#createSender()
+ */
+public class EventHubSender implements AutoCloseable {
+    private static final String EMPTY_PARTITION_KEY = "";
+
+    private final ServiceLogger logger = new ServiceLogger(EventHubSender.class);
+    private final int maxMessageSize;
+
+    /**
+     * Creates a new instance of the EventHubSender.
+     */
+    EventHubSender() {
+        this(ClientConstants.MAX_MESSAGE_LENGTH_BYTES);
+    }
+
+    /**
+     * Creates a new instance of this object with batches that are {@code maxMessageSize}.
+     *
+     * @param maxMessageSize Message size for each batch.
+     */
+    EventHubSender(int maxMessageSize) {
+        this.maxMessageSize = maxMessageSize;
+    }
+
+    /**
+     * Sends the {@code events} to the Event Hubs service. If {@link EventData#partitionKey()} is specified, the events
+     * are grouped by that key and sent to the service in batches.
+     *
+     * @param events Events to send to the service.
+     * @return A {@link Mono} that completes when all events are pushed to the service.
+     */
+    public Mono<Void> send(EventData... events) {
+        if (events == null || events.length == 0) {
+            return Mono.empty();
+        }
+
+        return send(Flux.just(events));
+    }
+
+    /**
+     * Sends the {@code events} to the specified Event Hubs {@code partitionId}. If {@link EventData#partitionKey()} is
+     * specified, these are dropped from being sent.
```

Comment:
I wonder if they should be dropped from being sent, or whether they should be sent in a separate batch (one extra batch per partition key)?

Reply:
I've updated it so that if the user calls `send(partitionId, EventData... events)`, if the `partitionKey` is specified, we'll ignore it (presumably a user error) and send the events to the given `partitionId`.

Reply:
Do you think this is correct, or should we send separate batches for each unique partition key / id?